[RzIL] Add concat, append handlers etc.

ogre2007 opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

I've just added concat IL opcode handler, and fixed mistake in the definition of the concat bv operation. If this is ok, I'll add another opcodes later

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

linked to #1918

ogre2007 committed:
Add and register concat il handler & fix mistake in the definition of concat operation

ogre2007 committed:
Actual fix in concat operation

ogre2007 committed:
Actual fix in concat operation

ogre2007 committed:
Edit vm_layer handlers table

ogre2007 committed:
Edit vm_layer handlers table

wargio commented:
looks like you forgot to change the unit test

XVilka commented:
Please also rebase.

thestr4ng3r commented on librz/include/rz_il/rzil_opcodes.h:
```diff
@@ -118,6 +118,16 @@ struct rzil_op_cast_t {
+
+/**
+ *  \struct rzil_op_concat_t
+ *  \brief op structure for concating bitv
```
This does not add any value over the name the op already has, please describe exactly what it does, for example from BAP documentation:
> Appends x and y so that in the resulting vector the vector x occupies the most significant part and y the least significant part.

XVilka replied:
Yes, append operates on two bitvectors, while concat operates on the list of bitvectors.

thestr4ng3r reviewed (changes requested):
It should be called "append" to correspond to BAP: http://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/Theory/module-type-Minimal/index.html#val-append

ogre2007 committed:
Merge branch 'il-bitv-append-concat' of https://github.com/ogre2007/rizin into il-bitv-append-concat

ogre2007 committed:
Renamed concat to append, bcs it should operate on 2 bitvs

wargio reviewed (changes requested):
unit test needs to be changed (https://github.com/rizinorg/rizin/blob/dev/test/unit/test_il_definitions.c) and also you need to add `void *rz_il_handler_concat(RzILVM *vm, RzILOp *op, RzILOpArgType *type);` definition in vm_layer.c

ogre2007 committed:
Added concat operation and opcode definition

ogre2007 commented:
I've renamed concat into append, and added new concat operation, but there is not RzList ARG type in opcodes, so its not correct currently. Should I add new ARG_TYPE?

wargio commented on librz/include/rz_il/rzil_vm.h:
```diff
@@ -100,6 +100,54 @@ RZ_API RzILEffect *rz_il_evaluate_effect(RzILVM *vm, RzILOp *op, RzILOpArgType *
+void *rz_il_handler_lsb(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_ule(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_sle(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_neg(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_not(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_add(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_sub(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_mul(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_div(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_sdiv(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_mod(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_smod(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_shiftl(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_shiftr(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_logical_and(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_logical_or(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_logical_xor(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+void *rz_il_handler_b0(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_b1(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_and_(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_or_(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_inv(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_cast(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_append(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_concat(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+
+void *rz_il_handler_perform(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_set(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_jmp(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_goto(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_seq(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_branch(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+void *rz_il_handler_load(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_store(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+//TODO: remove me when all the handlers are implemented
+void *rz_il_handler_unimplemented(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
```
i have removed them for a reason. just add the concat/append at the beginning of vm_layer.c

ogre2007 replied:
Sure! But the question is how to provide RzList as argument for concat inside the operation object, bcs there is no argtype for it

wargio replied:
i think you need to edit the RzILOpConcat structure to have an RzList, but i would wait for @XVilka for his input. i'm not sure if that is the right way to implement this.

wargio commented on librz/il/definitions/bitvector.c:
```diff
@@ -251,24 +251,48 @@ RZ_API RZ_OWN RzILBitVector *rz_il_bv_cut_tail(RZ_NONNULL RzILBitVector *bv, ut3
 	return ret;
 }
 
+/**
+ * Concat bitvectors from list bvs to get new bitvector
+ * \param bvs RzList of bitvectors
+ * \return ret RzILBitVector, the new bitvector
+ */
+RZ_API RZ_OWN RzILBitVector *rz_il_bv_concat(RZ_NONNULL RzList *bvs) {
+	rz_return_val_if_fail(bvs, NULL);
+	RzListIter *iter = bvs->head;
+	RzILBitVector *tmp = NULL, *ret = NULL, *bv = NULL;
+	rz_list_foreach(bvs, iter, bv){
+		if (iter == bvs->head){
+			ret = rz_il_bv_new(bv->len);
+			rz_il_bv_copy_nbits(bv, 0, ret, 0, bv->len);
+			tmp = ret;
+			continue;
+		}
+		ret = rz_il_bv_append(ret, bv);
+		rz_il_bv_free(tmp);
+		tmp = ret;
+	}
+	tmp = NULL;
+	return ret;
+}
```
i'm not sure about this. @XVilka

ogre2007 replied:
It actually works though. I can do it w/o append function, but it takes two foreach loops: 1st for collecting actual lengths, then we can allocate resulting bv, then we copy bvs content in the second foreach loop.

wargio replied:
looks correct, but the problem is if this is actually needed, or not. so this is why i asked to wait for him

XVilka commented:
On the one hand, such an operation could be emulated by calling multiple `append` operations. On the other hand, it might become handy for uplifting more complex instructions. I guess we can add a new `list` type for the arguments. @thestr4ng3r your thoughts?

thestr4ng3r commented:
From what I can tell, `RzILOpArgType` is only needed when an op **returns** a list, but not if it takes one. Or are there ops in BAP that can return lists of bitvectors? That would seem overcomplicated to me.
Regardless, I would just omit the concat completely for now. `RzList` is a linked list anyway and chaining multiple append ops would actually be one too.

ogre2007 committed:
Delete concat bv operation and opcode definitions, formatted

ogre2007 committed:
Format issues again

XVilka commented on librz/il/definitions/bitvector.c:
```diff
@@ -251,12 +251,12 @@ RZ_API RZ_OWN RzILBitVector *rz_il_bv_cut_tail(RZ_NONNULL RzILBitVector *bv, ut3
 }
 
 /**
- * Concat bv1 and bv2 to get new bitvector
```
Append then

XVilka commented on test/unit/test_il_definitions.c:
```diff
@@ -179,6 +179,7 @@ bool test_rzil_bv_algorithm(void) {
 	RzILBitVector *x, *y;
 	RzILBitVector *result;
 	RzILBitVector *add, *sub, *mul, *div, *mod;
+
```
Unnecessary change

XVilka commented:
@ogre2007 please rebase and address my comments, we can merge it then.

ogre2007 commented:
ok, Ill do it today

ogre2007 committed:
Merge branch 'rizinorg:dev' into il-bitv-append-concat

ogre2007 committed:
Minor fix

wargio commented on librz/il/vm_layer.c:
```diff
@@ -177,8 +177,8 @@ RZ_API bool rz_il_vm_init(RzILVM *vm, ut64 start_addr, ut32 addr_size, ut32 data
-	vm->op_handler_table[RZIL_OP_CONCAT] = &rz_il_handler_unimplemented; // &rz_il_handler_concat;
-	vm->op_handler_table[RZIL_OP_APPEND] = &rz_il_handler_unimplemented; // &rz_il_handler_append;
+	vm->op_handler_table[RZIL_OP_CONCAT] = &rz_il_handler_unimplemented; // &rz_il_handler_concat
+	vm->op_handler_table[RZIL_OP_APPEND] = &rz_il_handler_append; // &rz_il_handler_append;
```
```suggestion
	vm->op_handler_table[RZIL_OP_APPEND] = &rz_il_handler_append;
```

wargio commented on librz/include/rz_il/rzil_vm.h:
```diff
@@ -100,6 +100,52 @@ RZ_API RzILEffect *rz_il_evaluate_effect(RzILVM *vm, RzILOp *op, RzILOpArgType *
+void *rz_il_handler_msb(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_lsb(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_ule(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_sle(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_neg(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_not(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_add(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_sub(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_mul(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_div(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_sdiv(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_mod(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_smod(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_shiftl(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_shiftr(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_logical_and(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_logical_or(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_logical_xor(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+void *rz_il_handler_b0(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_b1(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_and_(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_or_(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_inv(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_cast(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_append(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+void *rz_il_handler_perform(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_set(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_jmp(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_goto(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_seq(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_branch(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+void *rz_il_handler_load(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+void *rz_il_handler_store(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
+//TODO: remove me when all the handlers are implemented
+void *rz_il_handler_unimplemented(RzILVM *vm, RzILOp *op, RzILOpArgType *type);
+
```
```suggestion
```

wargio replied:
please add the definition (`void *rz_il_handler_append(RzILVM *vm, RzILOp *op, RzILOpArgType *type);`) in vm_layer.c

This header mush **NOT** have any of these definitions there.

wargio committed:
Merge branch 'dev' into il-bitv-append-concat

wargio committed:
Added rz_il_handler_append to list

wargio committed:
Removed handlers from header
